### ISSUE TITLE
How to configure the boot loader from the SD card?

### ISSUE THREAD

**Issue by eproxus** [grisp/grisp2-rtems-toolchain#18]
Is it possible to tweak the boot loader via a configuration file on the SD-card somehow? E.g. point out a path to another `.zImage` file than the default `./zImage`? Can't find anything about configuration files related to Barebox...

**Comment by eproxus** [comment]
Digging into this now, perhaps it can solve this: https://barebox.org/doc/latest/user/booting-linux.html#bootloader-spec

**Comment by eproxus** [comment]
Ok, custom entries works with the following example in e.g. `/loader/entries/grisp.conf` on the SD-card:

```
title Erlang
linux /zImage
```

However, this requires removing the default settings in `/env/boot/mmc` which are:

```
#!/bin/sh

global.bootm.image="/mnt/mmc/zImage"
global.bootm.oftree="/mnt/mmc/oftree"
global.linux.bootargs.dyn.root="root=/dev/mmcblk0p2 rootflags='data=journal'"
```

Those seem to override the medium `/loader/*` files. Can we remove these defaults or change them to be backup somehow?

**Comment by eproxus** [comment]
Ok, custom entries works with the following example in e.g. `/loader/entries/grisp.conf` on the SD-card:

```
title Erlang
linux /zImage
```

However, the default settings in `/env/boot/mmc` conflicts with this somehow:

```
#!/bin/sh

global.bootm.image="/mnt/mmc/zImage"
global.bootm.oftree="/mnt/mmc/oftree"
global.linux.bootargs.dyn.root="root=/dev/mmcblk0p2 rootflags='data=journal'"
```

If I create a differently name image (e.g. `/beam.zImage`) and create a corresponding config file as above, the default boot loader configuration fails to boot. Once at the barebox shell I can type `boot mmc0` to boot from the SD card anyway and then the configuration is picked up. Can we change this default somehow?

**Comment by c-mauderer** [comment]
I think you can change the boot order by changing `/env/init/bootsource`. Maybe with your configh you would have to adapt that too. But I have very few experience with barebox. I would have preffered an U-Boot but Phytec doesn't provide one for this module.

Also note: We already have a custom build. So if you want to change the defaults, it would be possible in the sources of the bootloader too.

**Comment by eproxus** [comment]
Would it be possible to add the `mmc0` as a default before `mmc` to search the SD card first and then in case no loader specs are found, the default `zImage` from `mmc` would be used? Or is that a bad idea?

I'm not sure about the difference between `mmc` and `mmc0` in this case and why it works the way it does...

**Comment by peerst** [comment]
To be taken into account have two mmc devices:  the eMMC and the SD card.

We should find a solution where the boot order requirements we might have with this extra complication is met.  Unfortunately there both priorities of the two mmc devices make sense in certain use cases:

eMMC higher than SD card:   we might not want to boot from a SD card which is used for data storage

SD card higher than eMMC:   we might be able to override the eMMC boot in development or for unbricking a device in the field

Probably eMMC having a higher priority is the more common scenario

**Comment by c-mauderer** [comment]
> SD card higher than eMMC: we might be able to override the eMMC boot in development or for unbricking a device in the field

Note that the bootloader will always load from the primary medium. For the prototypes that's the NAND. For the new selected module it will be the eMMC on the module. For recovering a bootloader the serial download mode is necessary (or JTAG if someone finds out how to init the SDRAM via JTAG - I had no luck with this).

But assuming the bootloader on eMMC still works: You could think about using one of the switches to switch between boot sources. It's something that someone would have to implement and it most likely needs some C-Code.

Maybe detecting a file ("forceSDBoot") on the SD could be an alternative that would be doable with environment variables.

For the default environment: Think about which you want. We can then add it in barebox in the `arch/arm/boards/phytec-som-imx6/defaultenv-physom-imx6ull-grisp2` directory to have it on all produced boards.

**Comment by GalaxyGorilla** [comment]
@c-mauderer I'm currently playing around with Grisp2 (with EMMC) and your toolchain branch. As explained by @eproxus we use the FDT from a custom place and hence there is some special configuration for barebox (in `/loader/entries/grisp.conf`). I get this error here when I attempt to use an SD card filled with a grisp app.

```
barebox@PHYTEC phyCORE-i.MX6 ULL SOM with eMMC:/ boot mmc0
blspec: ignoring entry with incompatible devicetree "embeddedbrains,grisp2"
ext4 ext40: EXT2 rev 1, inode_size 256, descriptor size 32
Nothing bootable found on 'mmc0'
Nothing bootable found
```

I know you are not familiar with barebox, but does this ring a bell? Where can I find the `compatible` list for the board?

**Comment by c-mauderer** [comment]
I'm not really sure why the fdt "compatible" is checked during a boot command. I have seen that only during boot loader update where you could avoid it by executing the new boot loader from RAM before updating.

Anyway: There are two locations where the compatible string is set in the FDT:
1. In barebox itself: https://github.com/grisp/barebox/blob/e2d1530f2c5f4bf6fcf735480a511adc495f4630/arch/arm/dts/imx6ull-grisp2.dts#L15 or https://github.com/grisp/barebox/blob/3322880d50b964c17021f9a6089d191b6cf1586b/arch/arm/dts/imx6ull-phytec-phycore-som-emmc.dts#L16 for the new branch. That FDT is used for the boot loader itself.
2. In the FDT that is used for RTEMS: https://github.com/grisp/grisp2-rtems-toolchain/blob/cm/PCL-063-005/fdt/imx6ull-grisp2.dts#L14

For the new module I switched back to an upstream barebox because no adaptions have been necessary any more and using upstream makes the maintenance simpler. But most likely that is the reason for your new problem. Now the FDT used by barebox tells that it is compatible with `"phytec,imx6ul-pcl063-emmc", "fsl,imx6ull"` while the RTEMS FDT is compatible with `"embeddedbrains,grisp2", "phytec,imx6ul-pbacd-10", "phytec,imx6ul-pcl063", "fsl,imx6ul"`. Seems that I forgot to adapt the line in the FDT. Can you try changing
```
compatible = "embeddedbrains,grisp2", "phytec,imx6ul-pbacd-10", "phytec,imx6ul-pcl063", "fsl,imx6ul";
```
to
```
compatible = "embeddedbrains,grisp2", "phytec,imx6ul-pcl063-emmc", "fsl,imx6ull";
```
in the `grisp2-rtems-toolchain/fdt/imx6ull-grisp2.dts` or whichever devicetree file you use and re-build the fdt?

**Comment by GalaxyGorilla** [comment]
@c-mauderer I just tested with your fix and I can boot to an Erlang shell 👍 

Thanks!

**Comment by GalaxyGorilla** [comment]
@c-mauderer I'm currently trying to run an app within the emmc and was following your instructions from the README on your branch. However, with

```
8. cp grisp.img /dev/mmc1
```

I killed the bootloader, the board doesn't react anymore, just a red power light (I think). Any idea what happened here?

I'll try to recover the bootloader now following the README ...

**Comment by c-mauderer** [comment]
How did you create the eMMC image? Did you remember to add the Boot loader? That would be the second dd command in
https://github.com/grisp/grisp2-rtems-toolchain/blob/cm/PCL-063-005/README.md#emmc-image

**Comment by GalaxyGorilla** [comment]
> How did you create the eMMC image? Did you remember to add the Boot loader? That would be the second dd command in
> https://github.com/grisp/grisp2-rtems-toolchain/blob/cm/PCL-063-005/README.md#emmc-image

Yea, very likely I did this wrong last week :(
On the good side, this way I also test the bootloader recovery in the README ;)).

**Comment by peerst** [comment]
Hi Christian,

do I understand this right that the bootloader is together with the RTEMS image on the eMMC?

Is there a danger to brick a board when something goes wrong with this process?

Cheers,
-- Peer

**Comment by c-mauderer** [comment]
> On the good side, this way I also test the bootloader recovery in the README ;)).

Let me know if something does not work. I think I wrote the chapter in the README for the future series boards that will have some additional (unpopulated) jumpers. But if the bootloader from eMMC doesn't work, the board should fall back to a serial downloader anyway (at least I hope that).

> do I understand this right that the bootloader is together with the RTEMS image on the eMMC?

With the eMMC on the module, we don't have a lot of other memories to boot from. Therefore the bootloader is now on the eMMC before the first partition starts.

> Is there a danger to brick a board when something goes wrong with this process?

When writing the whole eMMC and not only data in a partition: Yes there is. But we have a recovery process using the serial downloader documented in the README.

**Comment by peerst** [comment]
Hi Christian

> On 30. Nov 2020, at 14:52, Christian Mauderer <notifications@github.com> wrote:
> 
> Is there a danger to brick a board when something goes wrong with this process?
> 
> When writing the whole eMMC and not only data in a partition: Yes there is. But we have a recovery process using the serial downloader documented in the README.
> 
> 

Ok, so when is that whole eMMC writing necessary?  Every time we want to write a RTEMS image on the eMMC?

Just wanted to know how we can make the normal software development process of overwriting the RTEMS image and the associated files as use friendly as possible (a chance of bricking is not so user friendly).  If its unavoidable we need to think internally about how we can make the process as risk free as possible.

Cheers,
-- Peer

**Comment by c-mauderer** [comment]
> Ok, so when is that whole eMMC writing necessary? Every time we want to write a RTEMS image on the eMMC?

To update RTEMS you can replace a file in the FAT file system. I can currently think of three reasons for replacing the eMMC image:
1. Something bad happened to the file system. In that case maybe it's possible to re-write only a part with the boot loader. But I haven't tried something like that yet.
2. If you want to replace the boot loader.
3. If you want to re-partition the eMMC.

**Comment by peerst** [comment]
So this eMMC image that you mentioned before which includes the bootloader is only necessary because the eMMC is not partitioned yet?

**Comment by c-mauderer** [comment]
I'm quite sure that it had some partitions when I delivered it. Most likely @GalaxyGorilla wanted to either have different partitions (bigger ones?) or wanted to try every part of the README which is a good idea. It's quite clear now that we have to add a big warning sign to the README there ;-)

Yes, from my point of view, re-writing the whole eMMC shouldn't be the normal use case. We have to make sure that we do a proper partitioning during the manufacturing tests.

**Comment by peerst** [comment]
I agree that the whole process needs to be tested

**Comment by GalaxyGorilla** [comment]
> I'm quite sure that it had some partitions when I delivered it. Most likely @GalaxyGorilla wanted to either have different partitions (bigger ones?) or wanted to try every part of the README which is a good idea. It's quite clear now that we have to add a big warning sign to the README there ;-)

I just wanted to get an Erlang node running on the eMMC, that's all. And the README talks about writing images to the eMMC so I wanted to follow that path :). Sorry for the mess, I guess that qualifies me as DAU ("dümmster anzunehmender user") tester 😂 .

**Comment by c-mauderer** [comment]
> > I'm quite sure that it had some partitions when I delivered it. Most likely @GalaxyGorilla wanted to either have different partitions (bigger ones?) or wanted to try every part of the README which is a good idea. It's quite clear now that we have to add a big warning sign to the README there ;-)
> 
> I just wanted to get an Erlang node running on the eMMC, that's all. And the README talks about writing images to the eMMC so I wanted to follow that path :). Sorry for the mess, I guess that qualifies me as DAU ("dümmster anzunehmender user") tester joy .

Not at all a problem. It shows that the README isn't clear enough. Do you want to extend it with your experiences or should I try to make it more clear?

**Comment by GalaxyGorilla** [comment]
> Not at all a problem. It shows that the README isn't clear enough. Do you want to extend it with your experiences or should I try to make it more clear?

Don't worry, I'll provide a PR later after the board runs, you can check my changes then.

**Comment by eproxus** [comment]
Some comments after testing with the latest hardware:

1. Booting from SD no longer works:
   ```
   barebox@Phytec phyCORE-i.MX6 ULL SOM with eMMC:/ boot mmc0 
   Nothing bootable found on 'mmc0'
   Nothing bootable found
   barebox@Phytec phyCORE-i.MX6 ULL SOM with eMMC:/ cat /mnt/mmc0.0/loader/entries/erlang.conf
   title       robot 0.1.0
   linux       /robot/erts-11.2.2.4/bin/beam.zImage
   devicetree  /robot/erts-11.2.2.4/bin/imx6ull-grisp2.dtb
   ```
   This used to work before (with these exact files).
2. Should we prefer the SD card over the eMMC by default on the shipped boards? In 99% of the cases users will try to deploy new software to an SD card and not the eMMC, right?

**Comment by peerst** [comment]
> 2. Should we prefer the SD card over the eMMC by default on the shipped boards? In 99% of the cases users will try to deploy new software to an SD card and not the eMMC, right?

But then out of the box nothing would boot anymore since the initial image is on eMMC.

Can we have a boot sequence of first try SD and if nothing bootable there go to eMMC?

**Comment by eproxus** [comment]
> But then out of the box nothing would boot anymore since the initial image is on eMMC.

It would boot eMMC if booting from SD card fails.

**Comment by eproxus** [comment]
@c-mauderer Regarding my comment (1): Have we downgraded Barebox from 2020.07 to 2019.01? Could this be the issue? See https://github.com/grisp/grisp2-rtems-toolchain/commit/b040537ca35ee52111aa67a789ad2f1de251f909#diff-2fa831a9501463be1c122c52104851b0d93e708323e08dd500918cadec2f359fL3-R3

**Comment by c-mauderer** [comment]
> 
> 
> @c-mauderer Regarding my comment (1): Have we downgraded Barebox from 2020.07 to 2019.01? Could this be the issue? See [b040537#diff-2fa831a9501463be1c122c52104851b0d93e708323e08dd500918cadec2f359fL3-R3](https://github.com/grisp/grisp2-rtems-toolchain/commit/b040537ca35ee52111aa67a789ad2f1de251f909#diff-2fa831a9501463be1c122c52104851b0d93e708323e08dd500918cadec2f359fL3-R3)

I did not touch the barebox that is delivered by Phytec for this hardware. The custom barebox was necessary when we had an eMMC on the base board but with the eMMC on the module, we can use the official version.

Yes, that is an older version. But that shouldn't change the boot order. It's more likely that the default config of the Phytec Barebox has a boot order that starts with eMMC followed by SD.

I think we can either change that by setting some Barebox configs during the test (I have no idea which one exactly so don't hesitate to tell me if you know it) or by building our own Barebox with the right default config and update the existing Barebox to that.

Note that this also depends on how you want to deliver your software. At the moment during the test I create a partition that just fills the whole space of the eMMC and then I unpack a tar file to it. If you want to update Barebox, I think it would be better to just write a complete image (or at least the first few megabytes - we don't have to write the whole empty space).

**Comment by GalaxyGorilla** [comment]
@c-mauderer to understand this situation a bit more: you're using this old version from phytec because it's the latest recommended image they provide, right?

Can't we just test and use upstream barebox? Tinkering with this old version already gives us problems with boot specs that we did not have before. And 2019.01 is quite old by now anyway. Are there hard reasons why to stick to the recommended one?

A quick look through barebox upstream and https://git.phytec.de/barebox/log/?h=v2019.01.0-phy shows that everything important regarding the emmc is upstream anyway.

**Comment by eproxus** [comment]
Fixed in #25.
